pkg/preflight/preflight.go

```diff
@@ -0,0 +1,295 @@
+	}
+	image := ds.Spec.Template.Spec.Containers[0].Image
+
+	p.log.Info("daemonset image for preflight validation", "image", image)
+
+	repo, digests, auth, err := p.registryAPI.GetLayersDigests(ctx, image)
+	if err != nil {
+		p.log.Info("Failed to get layers digests for image", "image", image, "err", err)
+		return false, nil
+	}
+
+	for i := len(digests) - 1; i >= 0; i-- {
+		layer, err := p.registryAPI.GetLayerByDigest(repo, digests[i], auth)
+		if err != nil {
+			p.log.Info("Failed to extract/access image", "image", image, "err", err)
+			return false, nil
+		}
+		dtk, err := p.registryAPI.ExtractToolkitRelease(layer)
+		if err != nil {
+			p.log.Info("dtk info not present", "image", image, "layerIndex", i)
+			continue
+		}
+
+		p.log.Info("dtk info present in layer", "layerIndex", i)
+		if dtk.KernelFullVersion != upgradeKernelVersion {
+			p.log.Info("DTK kernel version differs from the upgrade node version", "dtkVersion", dtk.KernelFullVersion, "upgradeVersion", upgradeKernelVersion)
+			return false, nil
+		}
+		return true, nil
+	}
+
+	p.log.Info("DTK info not present on any layer of the image, not good", "image", image)
+	return false, nil
+}
+
+func (p *preflight) PrepareRuntimeInfo(ctx context.Context, image string) (*runtime.RuntimeInformation, error) {
+	var err error
+	runInfo := p.runtimeAPI.InitRuntimeInfo()
+	runInfo.OperatingSystemMajor, runInfo.OperatingSystemMajorMinor, runInfo.OperatingSystemDecimal, err = p.getOSData(ctx, image)
+	if err != nil {
+		p.log.Error(err, "Failed to get os data for preflight")
+		return nil, err
+	}
+	runInfo.KernelFullVersion, err = p.getKernelFullVersion(ctx, image)
+	if err != nil {
+		p.log.Error(err, "Failed to get kernel full version for preflight")
+		return nil, err
+	}
+	runInfo.KernelPatchVersion, err = p.kernelAPI.PatchVersion(runInfo.KernelFullVersion)
+	if err != nil {
+		p.log.Error(err, "Failed to get kernel patch version for preflight", "fullVersion", runInfo.KernelPatchVersion)
+		return nil, err
+	}
+
+	runInfo.ClusterVersion, runInfo.ClusterVersionMajorMinor, err = p.clusterAPI.Version(ctx)
+	if err != nil {
+		p.log.Error(err, "Failed to get cluster version info for preflight")
+		return nil, err
+	}
+	// for preflight we do not need a real push secret, since preflight validation is not
+	// doing any image uploading
+	runInfo.PushSecretName = "dummyPushSecret"
+
+	runInfo.OSImageURL, err = p.clusterAPI.OSImageURL(ctx)
+	if err != nil {
+		p.log.Error(err, "failed to os image url for preflight")
+		return nil, err
+	}
+
+	// no need for real proxy configuration, since preflight check do not
+	// setup any object, just verifiying images
+	runInfo.Proxy = proxy.Configuration{}
+
+	return runInfo, nil
+}
+
+func (p *preflight) getOSData(ctx context.Context, image string) (string, string, string, error) {
+	layer, err := p.registryAPI.LastLayer(ctx, image)
+	if err != nil {
+		p.log.Error(err, "failed to get last layer of image", "image", image)
+		return "", "", "", err
```

Comment:
WDYT about returning the error messages instead of logging them?

Reply:
I prefer logging and returning the "virgin" error message with no wrappers around it.
Find it easier to track the root cause of an error.

When wrapping error, you may miss a few layers as it is wrapped in multiple places by multiple contributors but logging the error is easier to grep and you know the error is always the "root" error.

Reply:
i think we should do both, especially the first time we encounter the error

Reply:
I think doing both would be:
```suggestion
		p.log.Error(err, "failed to get last layer of image", "image", image)
		return "", "", "", fmt.Errorf("failed to get last layer of image %s: %v", image, err)
)
```
WDYT?